src/polycap-description.c

```diff
@@ -0,0 +1,438 @@
+			r = polycap_rng_uniform(rng);
+			src_rad_x = description->src_x * sqrt(fabs(r)); ////sqrt to simulate source intensity distribution (originally src_x * r/sqrt(r) )
+			r = polycap_rng_uniform(rng);
+			src_rad_y = description->src_y * sqrt(fabs(r)); ////sqrt to simulate source intensity distribution
+			r = polycap_rng_uniform(rng);
+			phi = 2.0*M_PI*fabs(r);
+			src_start_x = src_rad_x * cos(phi) + description->src_shiftx;
+			src_start_y = src_rad_y * sin(phi) + description->src_shifty;
+			if((description->src_sigx * description->src_sigy) < 1.e-20){ //uniform distribution over PC entrance
+				r = polycap_rng_uniform(rng);
+				pc_rad = description->profile->ext[0] * sqrt(fabs(r));
+				r = polycap_rng_uniform(rng);
+				phi = 2.0*M_PI*fabs(r);
+				pc_x = pc_rad * cos(phi) + start_coords.x;
+				pc_y = pc_rad * sin(phi) + start_coords.y;
+				start_direction.x = pc_x - src_start_x;
+				start_direction.y = pc_y - src_start_y;
+				start_direction.z = description->d_source;
+			} else { //non-uniform distribution, direction vector is within +- sigx
+				r = polycap_rng_uniform(rng);
+				start_direction.x = description->src_sigx * (1.-2.*fabs(r));
+				r = polycap_rng_uniform(rng);
+				start_direction.y = description->src_sigy * (1.-2.*fabs(r));
+				start_direction.z = 1.;
+			}
+			polycap_norm(&start_direction);
+printf("Here4; start(x,y,z):%f, %f, %f\n",start_direction.x,start_direction.y,start_direction.z);
+
+			// Create photon structure
+			photon = polycap_photon_new(rng, start_coords, start_direction, start_electric_vector, n_energies, energies);
+
+printf("Here5\n");
+			// Launch photon
+			istart++; //Here all photons that started, also enter the polycapillary
+			photon->i_refl = 0; //set reflections to 0
+			iesc = polycap_photon_launch(photon, description);
+			//if iesc == -1 here a new photon should be simulated/started.
+			//	essentially do j-1 as this will have same effect
+		} while(iesc == -1);
+
+		//COUNT function here... If required...
+
+printf("Here6\n");
+		if(thread_id == 0 && (double)i/((double)icount/(double)thread_cnt/10.) >= 1.){
+			printf("%d%%\t%" PRId64 "\t%f\n",((j*100)/(icount/thread_cnt)),photon->i_refl,photon->exit_coords.z);
+			i=0;
+		}
+		i++;//counter just to follow % completed
+
+		//save photon->weight in thread unique array
+		for(k=0; k<n_energies; k++) weights[k] += photon->weight[k];
+		#pragma omp critical
+		{
+		sum_irefl += photon->i_refl;
+		}
+printf("Here7\n");
+
+		//free photon structure (new one created for each for loop instance)
+		polycap_photon_free(photon);
+
+	} //for(j=0; j < icount; j++)
+
+	#pragma omp critical
+	{
+	sum_istart += istart;
+	for(i=0; i<n_energies; i++) sum_weights[i] += weights[i];
+	}
+	polycap_rng_free(rng);
+	free(weights);
+} //#pragma omp parallel
+printf("Here8\n");
+
+	printf("Average number of reflections: %f\n",(double)sum_irefl/icount);
+
+	*efficiencies = malloc(sizeof(double) * n_energies);
```

Comment:
Probeer iets met een tijdelijke variable zoals:

````c
double *efficiencies_temp = malloc(sizeof(double) * n_energies);
*efficiencies = efficiencies_temp; // do not free efficiencies_temp!!
for(i=0; i<n_energies; i++){
 +		efficiencies_temp[i] = (sum_weights[i] / (double)sum_istart) * description->open_area;
 +	}
````

Reply:
Yup, this works. Silly C :)

Enige probleem dat ik nu heb is dat:
- mijn uitkomsten compleet fout zijn (hij vindt nergens reflecties aan de wanden van eht capillair, dus daar gaat nog iets fout)
- dat thread_id getal is nog steeds heel erg foutief.
Valgrind zelf ziet niet echt iets vreemd.
Als je terug bent (kwestie dat ik je niet oneindig stoor op je reis) moet je het anders maar eens draaien, ik gebruik de ellip_l9.inp als input, maar dat thread_id gedoe heb je bij alles natuurlijk.
In tussentijd hou ik me dan bezig met wat andere taakjes die zijn blijven liggen :)

Reply:
Dit is dus waarom unit tests zo goed zijn: je test elke individuele functie door de verwachte resultaten te vergelijken met de eigenlijke 😄 

Morgen kan ik u niet van dienst zijn aangezien we voor een paar dagen weg gaan naar een andere stad. Ik zal geen of weinig internet toegang hebben...